### PR TITLE
feat(scripts) upload bundle script

### PR DESCRIPTION
**Pull request by loweisz**
Adds the upload script to `app-scripts` in interactive and in non-interactive (`--ci`) mode.

### CODE REVIEW

File: packages/@contentful--app-scripts/lib/upload/index.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+const { createAppBundle } = require('./create-app-bundle');
+
+const { buildAppUploadSettings } = require('./build-upload-settings');
+
+module.exports = {
+  async interactive() {
+    const settings = await buildAppUploadSettings();
+    await createAppBundle(settings);
+  },
+  async nonInteractive(program, options) {
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/index.js:11]: Why do we need `program` here?

File: packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+function validateArguments(options) {
+  Object.entries(requiredOptions).forEach(([option, argument]) => {
+    if (!options[option]) {
+      console.log(
+        `
+  ${chalk.red('Invalid Arguments:')} the argument ${chalk.cyan(argument)} was not defined
+  Run ${chalk.dim('contentful-app-scripts upload --help')} to see all required arguments
+        `
+      );
+      throw new Error('Invalid Arguments');
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js:24]: nit: this might be a `TypeError`

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+async function createZipFileFromDirectory(path) {
+  try {
+    let zip = new JSZip();
+    const filePaths = await new Promise((res) => {
+      glob(path + '/**/*', { nodir: true }, (err, paths) => {
+        res(paths);
+      });
+    });
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:35]: You can use `util.promisify` here to stick with the async/await style we have elsewhere

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+const fs = require('fs');
+const JSZip = require('jszip');
+const { showCreationError } = require('../utils');
+
+async function convertPathToFileData(filePath) {
+  return new Promise((res) => {
+    fs.readFile(filePath, 'utf8', (err, data) => {
+      res([filePath, data]);
+    });
+  });
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:12]: You can use `fs/promises` instead of wrapping. Although, being a script, also sync API would do

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      localPath = localPath.substring(localPath.length, path.length);
+    }
+    zip.file(localPath, data);
+  });
+  return zip;
+}
+
+async function createZipFileFromDirectory(path) {
+  try {
+    let zip = new JSZip();
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:30]: You can use [`adm-zip`](https://github.com/cthackers/adm-zip) here: it's MIT licensed and it allows adding a full folder by its path (so no need to do `fs` manipulation)

File: packages/@contentful--app-scripts/lib/upload/definition-api.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
+const chalk = require('chalk');
+const inquirer = require('inquirer');
+const ora = require('ora');
+
+async function fetchDefinitions(client, orgId) {
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/definition-api.js:5]: Why do we duplicate this?

File: packages/@contentful--app-scripts/lib/upload/organization-api.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+const inquirer = require('inquirer');
+const chalk = require('chalk');
+const ora = require('ora');
+
+async function fetchOrganizations(client) {
```
[the remark this answers is not in the document]
Comment (shikaan) [reply packages/@contentful--app-scripts/lib/upload/organization-api.js:5]: Also using the error wrapper I was suggesting below and the plain client in CMA, this can become a couple of lines of code. You might not need this utility in the first place

File: packages/@contentful--app-scripts/lib/upload/organization-api.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  const { organizationId } = await inquirer.prompt([
+    {
+      name: 'organizationId',
+      message: 'Select an organization for your app:',
+      type: 'list',
+      choices: organizations,
+    },
+  ]);
+  return organizations.find((org) => org.value === organizationId);
+}
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/organization-api.js:47]: Selecting from a list is something we do quite often. I would extract a utilty for it and reuse it for definitions and organizaions

File: packages/@contentful--app-scripts/lib/upload/definition-api.js (hunk 6 above)
Comment (shikaan) [reply packages/@contentful--app-scripts/lib/upload/definition-api.js:5]: Also, for the rest of the file, see comments on `organization-api`

File: packages/@contentful--app-scripts/lib/upload/index.js (hunk 1 above)
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/index.js:11]: Why `program` here?

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
+const glob = require('glob');
+const chalk = require('chalk');
+const fs = require('fs');
+const JSZip = require('jszip');
+const { showCreationError } = require('../utils');
+
+async function convertPathToFileData(filePath) {
+  return new Promise((res) => {
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:8]: You can use `fs/promises` instead of wrapping. Although, being a script, also sync API would do

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js (hunk 5 above)
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:30]: You can use [`adm-zip`](https://github.com/cthackers/adm-zip) here: it's MIT licensed and it allows adding a full folder by its path (so no need to do `fs` manipulation)

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    zip.file(localPath, data);
+  });
+  return zip;
+}
+
+async function createZipFileFromDirectory(path) {
+  try {
+    let zip = new JSZip();
+    const filePaths = await new Promise((res) => {
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:31]: You can use `util.promisify` here to stick with the async/await style we have elsewhere

**Review by shikaan — COMMENTED** [review]
It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)

One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?

File: packages/@contentful--app-scripts/lib/upload/organization-api.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/organization-api.js]: I actually already did that in the second PR about activation. But I moved it now here too

File: packages/@contentful--app-scripts/lib/upload/organization-api.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      value: org.sys.id,
+    }));
+  } catch (err) {
+    console.log(`
+${chalk.red(
+  'Error:'
+)} Could not fetch your organizations. Make sure you provided a valid access token.
+
+${err.message}
+    `);
```
[the remark this answers is not in the document]
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/organization-api.js]: Good idea 👍

File: packages/@contentful--app-scripts/lib/upload/index.js (hunk 1 above)
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/index.js]: I used it in a previous implementation, forgot to remove it

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js (hunk 3 above)
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js]: it's not used anymore anyway :D

File: packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js (hunk 5 above)
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js]: that was a really good tip 😬  Could remove half of the code haha

File: packages/@contentful--app-scripts/lib/upload/organization-api.js (hunk 8 above)
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/organization-api.js]: nice idea I abstracted it in utils 👍

**Comment by loweisz** [comment]
> It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)
> 
> One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?

@shikaan It actually will just ask for them and ignore the ones passed. I tried to make them conditional that when you pass `--ci` you will get a better notification that you have to pass them in that case. But somehow I couldn't find a good way 😬

But I think we could throw when passed arguments without `--ci`, do you think that would make more sense than just ignoring it?

**Comment by shikaan** [comment]
> > It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)
> > One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?
> 
> @shikaan It actually will just ask for them and ignore the ones passed. I tried to make them conditional that when you pass `--ci` you will get a better notification that you have to pass them in that case. But somehow I couldn't find a good way 😬
> 
> But I think we could throw when passed arguments without `--ci`, do you think that would make more sense than just ignoring it?

I was more thinking about we don't ask for the things we already know. So, if something is passed (even in interactive mode), we don't request further details on that.

TL;DR: we keep the same interface for interactive and non-interactive commands, but in the interactive case the arguments are all optional. WDYT?

**Comment by loweisz** [comment]
> > > It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)
> > > One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?
> > 
> > 
> > @shikaan It actually will just ask for them and ignore the ones passed. I tried to make them conditional that when you pass `--ci` you will get a better notification that you have to pass them in that case. But somehow I couldn't find a good way 😬
> > But I think we could throw when passed arguments without `--ci`, do you think that would make more sense than just ignoring it?
> 
> I was more thinking about we don't ask for the things we already know. So, if something is passed (even in interactive mode), we don't request further details on that.
> 
> TL;DR: we keep the same interface for interactive and non-interactive commands, but in the interactive case the arguments are all optional. WDYT?

Ahh I see ! That's a good idea, will look into it

File: packages/@contentful--app-scripts/lib/upload/create-app-bundle.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+const ora = require('ora');
+const { showCreationError } = require('../utils');
+const { createClient } = require('contentful-management');
+
+const { createAppUpload } = require('./create-app-upload');
+
+async function createAppBundleFromUpload(settings, appUploadId) {
+  const clientSpinner = ora('Verifing your upload').start();
+  const client = createClient({ accessToken: settings.accessToken });
+  const org = await client.getOrganization(settings.organisation.value);
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:11]: nit: organi**z**ation for consistency

**Review by shikaan — COMMENTED** [review]
Looks great 🎉  We only need to cover it with some tests and we're good to go 🚢

File: packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  token: '--token',
+};
+
+function validateArguments(options) {
+  Object.entries(requiredOptions).forEach(([option, argument]) => {
+    if (!options[option]) {
+      console.log(
+        `
+  ${chalk.red('Invalid Arguments:')} the argument ${chalk.cyan(argument)} was not defined
+  Run ${chalk.dim('contentful-app-scripts upload --help')} to see all required arguments
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js:21]: That command only works when the `contentful-app-scripts` are installed globally. If there is a chance that this error is displayed when using the CCA output, we should provide an alternative call here. Even `npx contentful-app-scripts --help` might do it.

Also: This should be `@contentful/app-scripts`

File: packages/@contentful--app-scripts/lib/upload/create-app-bundle.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  console.log(`
+  ${chalk.green(`NEXT STEPS:`)}
+
+    ${chalk.bold('You can activate this app bundle in your apps settings:')}
+
+      ${chalk.underline(`https://app.contentful.com/deeplink?link=app-definition-list`)}
+
+    ${chalk.bold('or by simply running the cli command:')}
+
+      ${chalk.underline(`contentful-app-scripts activate`)}
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:69]: Same issue with the app scripts package here.

File: packages/@contentful--app-scripts/bin/app-scripts [hunk 15]
```diff
@@ -3,32 +3,41 @@
...
+    })
 
-    return command();
-  })
+  program
+    .command('upload')
+    .description('Upload your build folder and create an app bundle')
+    .option('--bundle-dir [directory]', 'The directory of your build folder')
+    .option('--organization-id [orgId]', 'The id of your organisation')
+    .option('--definition-id  [defId]', 'The id of your apps definition')
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/bin/app-scripts:31]:
```suggestion
    .option('--definition-id [defId]', 'The id of your apps definition')
```

File: packages/@contentful--app-scripts/bin/app-scripts [hunk 16]
```diff
@@ -3,32 +3,41 @@
 const { program } = require('commander');
 const { version } = require('../package.json');
 
-const {createAppDefinition} = require('../')
+const {createAppDefinition, upload} = require('../')
 
-function getCommand(program, command) {
+
+function runCommand(command, options) {
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/bin/app-scripts:9]: This must be async and await for `nonInteractive()` and `interactive()`

File: packages/@contentful--app-scripts/bin/app-scripts [hunk 17]
```diff
@@ -3,32 +3,41 @@
...
   program
-  .command('create-app-definition')
-  .description('Create a new AppDefinition for an App')
-  .action(() => {
-    const command = getCommand(program, createAppDefinition);
+    .command('create-app-definition')
+    .description('Create a new AppDefinition for an App')
+    .action((options) => {
+      runCommand(createAppDefinition, options);
+    })
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/bin/app-scripts:24]: Should be an async function with await

File: packages/@contentful--app-scripts/bin/app-scripts [hunk 18]
```diff
@@ -3,32 +3,41 @@
...
+    .command('upload')
+    .description('Upload your build folder and create an app bundle')
+    .option('--bundle-dir [directory]', 'The directory of your build folder')
+    .option('--organization-id [orgId]', 'The id of your organisation')
+    .option('--definition-id  [defId]', 'The id of your apps definition')
+    .option('--token [accessToken]', 'Your content management access token')
+    .option('--comment [comment]', 'Optional comment for the created bundle')
+    .action((options) => {
+      runCommand(upload, options);
+    })
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/bin/app-scripts:36]: Should be an async function with await

**Review by andipaetzold — CHANGES_REQUESTED** [review]
You alternate between using `module.exports = ` and `exports.xxx`. Any reason for that?
I'd make this consistent and always use `module.exports`

Also, whether a log line has a trailing period `.` is inconsistent.

File: packages/@contentful--app-scripts/lib/upload/create-app-bundle.js [hunk 19]
```diff
@@ -9,8 +9,8 @@ async function createAppBundleFromUpload(settings, appUploadId) {
   console.log(settings.definition);
   const clientSpinner = ora('Verifing your upload').start();
   const client = createClient({ accessToken: settings.accessToken });
-  const org = await client.getOrganization(settings.organisation.value);
-  const appDefinition = await org.getAppDefinition(settings.definition.value);
+  const organization = await client.getOrganization(settings.organisation.value);
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:12]: Manuel was talking about `organisation` with an `s` here.

**Comment by loweisz** [comment]
> You alternate between using `module.exports = ` and `exports.xxx`. Any reason for that?
> I'd make this consistent and always use `module.exports`
> 
> Also, whether a log line has a trailing period `.` is inconsistent.

Not really `exports` was used in the code base before. I changed it to `module.exports` now everywhere and also added a `.` to all logs 👍

File: packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js (hunk 13 above)
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js]: True ! I was planning to change that with the real command, changed it now

Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js]: True ! I was planning to change that with the real command, changed it now

File: packages/@contentful--app-scripts/lib/upload/create-app-bundle.js [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+  }
+  bundleSpinner.stop();
+  return appBundle;
+}
+
+async function createAppBundleFromSettings(settings) {
+  let appUpload = null;
+  try {
+    appUpload = await createAppUpload(settings);
+    console.log('UPLOAD', appUpload);
```
Comment (shikaan) [inline packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:34]: Is this leftover? 🤔

File: packages/@contentful--app-scripts/bin/app-scripts [hunk 21]
```diff
@@ -3,32 +3,41 @@
...
   program
-  .command('create-app-definition')
-  .description('Create a new AppDefinition for an App')
-  .action(() => {
-    const command = getCommand(program, createAppDefinition);
+    .command('create-app-definition')
+    .description('Create a new AppDefinition for an App')
+    .action(async (options) => {
+      await runCommand(createAppDefinition, options);
+    })
```
Comment (shikaan) [inline packages/@contentful--app-scripts/bin/app-scripts:24]: super-nit

```suggestion
    .action((options) => runCommand(createAppDefinition, options))
```

File: packages/@contentful--app-scripts/bin/app-scripts [hunk 22]
```diff
@@ -3,32 +3,41 @@
...
-const {createAppDefinition} = require('../')
+const {createAppDefinition, upload} = require('../')
 
-function getCommand(program, command) {
+
+async function runCommand(command, options) {
   const {ci} = program.opts()
-  return ci ? command.nonInteractive : command.interactive;
+  return ci ? await command.nonInteractive(options) : await command.interactive(options);
 }
```
Comment (shikaan) [inline packages/@contentful--app-scripts/bin/app-scripts:12]: nit: I think this was good as it was: you're returning whatever the command returns, no need to `await`

**Review by shikaan — APPROVED** [review]
Great job 🎉 

There is only one - possible - leftover and a couple of `nit`s (mostly for code hygiene, no blockers)

File: packages/@contentful--app-scripts/bin/app-scripts (hunk 22 above)
Comment (andipaetzold) [reply packages/@contentful--app-scripts/bin/app-scripts:12]: I always prefer adding await on every level. That way we get a proper stack trace. The outcome when running the code will be the same.

Comment (loweisz) [reply packages/@contentful--app-scripts/bin/app-scripts:12]: I think in the end it doesn't really matter when in both cases `runCommand` will return a promise, or am I missing something ? @andipaetzold what was your intention to use await here?

File: packages/@contentful--app-scripts/lib/create-app-definition/create-app-definition.js [hunk 23]
```diff
@@ -36,18 +36,24 @@ function assertValidArguments(accessToken, appDefinitionSettings) {
   if (!isPlainObject(appDefinitionSettings) || !has(appDefinitionSettings, 'locations')) {
     throwValidationException(
       'AppDefinitionSettings',
-      `Expected plain object with 'location' property, got ${JSON.stringify(appDefinitionSettings, null, 2)}`
-        `Example: ${JSON.stringify({
+      `Expected plain object with 'location' property, got ${JSON.stringify(
+        appDefinitionSettings,
+        null,
+        2
+      )}``Example: ${JSON.stringify(
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/lib/create-app-definition/create-app-definition.js:43]: Does this syntax with double ` work?

File: packages/@contentful--app-scripts/lib/upload/create-app-bundle.js [hunk 24]
```diff
@@ -64,9 +66,12 @@ async function createAppBundle(settings) {
 
     ${chalk.bold('or by simply running the cli command:')}
 
-      ${chalk.underline(`contentful-app-scripts activate`)}
+      ${chalk.underline(`'npx @contentful/app-scripts activate'`)}
```
Comment (andipaetzold) [inline packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:69]:
```suggestion
      ${chalk.underline('npx @contentful/app-scripts activate')}
```

File: packages/@contentful--app-scripts/lib/upload/create-app-bundle.js (hunk 19 above)
Comment (andipaetzold) [reply packages/@contentful--app-scripts/lib/upload/create-app-bundle.js]: The other way round. We use American English in contentful. So: `organization`.
Sorry for being unspecific.

File: packages/@contentful--app-scripts/lib/create-app-definition/create-app-definition.js (hunk 23 above)
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/create-app-definition/create-app-definition.js:43]: I actually haven't changed something here it looks like it came from prettier when I changed something else in the file, so it probably will work when it worked before :D

Comment (andipaetzold) [reply packages/@contentful--app-scripts/lib/create-app-definition/create-app-definition.js:43]: Well it was a double ` before, so it didn't change. But e.g. `console.log(`test``test`)` throws an error for me as that's the syntax for template literals: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Comment (loweisz) [reply packages/@contentful--app-scripts/lib/create-app-definition/create-app-definition.js]: Yes you are right, I think there was a `,` lost as `throwValidationException` is also accepts three arguments. 
I added it now that the example is the `details` in the function

File: packages/@contentful--app-scripts/lib/upload/create-app-bundle.js (hunk 19 above)
Comment (loweisz) [reply packages/@contentful--app-scripts/lib/upload/create-app-bundle.js]: Ahhh I see, okay I changed it now everywhere to `organization`

**Review by andipaetzold — APPROVED** [review]
